src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -0,0 +1,1664 @@
+  C_WriteRequest(CephContext *cct, ImageCache::Extents &&image_extents,
+		 bufferlist&& bl, int fadvise_flags, Context *user_req)
+    : C_GuardedBlockIORequest(cct), m_cct(cct), image_extents(std::move(image_extents)),
+      bl(std::move(bl)), fadvise_flags(fadvise_flags),
+      user_req(user_req), _on_finish(NULL), m_user_req_completed(false) {
+    ldout(m_cct, 99) << this << dendl;
+  }
+
+  ~C_WriteRequest() {
+    ldout(m_cct, 99) << this << dendl;
+  }
+
+  void complete_user_request(int r) {
+    bool initial = false;
+    if (m_user_req_completed.compare_exchange_strong(initial, true)) {
+      ldout(m_cct, 15) << this << " completing user req" << dendl;
+      user_req->complete(r);
+    } else {
+      ldout(m_cct, 20) << this << " user req already completed" << dendl;
+    }
+  }
+
+  virtual void send() override {
+    /* Should never be called */
+    ldout(m_cct, 6) << this << " unexpected" << dendl;
+  }
+  
+  virtual void finish(int r) {
+    ldout(m_cct, 6) << this << dendl;
+
+    complete_user_request(r);
+    _on_finish->complete(0);
+  }
+
+  virtual const char *get_name() const override {
+    return "C_WriteRequest";
+  }  
+};
+
+/*
+ * Performs the log event append operation for all of the scheduled
+ * events.
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::append_scheduled_ops(void)
+{
+  WriteLogOperations ops;
+  int append_result = 0;
+
+  {
+    Mutex::Locker locker(m_lock);
+    if (m_ops_to_append.empty()) {
+      return;
+    }
+  }
+  {
+    Mutex::Locker locker(m_log_append_lock);
+
+    {
+      Mutex::Locker locker(m_lock);
+      std::swap(ops, m_ops_to_append);
+    }
+    
+    if (ops.size()) {
+      alloc_op_log_entries(ops);
+      append_result = append_op_log_entries(ops);
+    }
+  }
+
+  if (ops.size()) {
+    complete_op_log_entries(ops, append_result);
+  }
+}
+
+/*
+ * Takes custody of ops. They'll all get their log entries appended,
+ * and have their on_write_persist contexts completed once they and
+ * all prior log entries are persisted everywhere.
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::schedule_append(WriteLogOperations &ops)
+{
+  {
+    Mutex::Locker locker(m_lock);
+
+    m_ops_to_append.splice(m_ops_to_append.end(), ops);
+  }
+
+  m_log_append_finisher.queue(new FunctionContext([this](int r) {
+	append_scheduled_ops();
+      }));
+}
+
+/*
+ * Performs the pmem buffer flush on all scheduled ops, then schedules
+ * the log event append operation for all of them.
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::flush_then_append_scheduled_ops(void)
+{
+  WriteLogOperations ops;
+  {
+    Mutex::Locker locker(m_lock);
+    std::swap(ops, m_ops_to_flush);
+  }
+
+  /* Ops subsequently scheduled for flush may finish before these,
+   * which is fine. We're unconcerned with completion order until we
+   * get to the log message append step. */
+  if (ops.size()) {
+    flush_pmem_buffer(ops);
+    schedule_append(ops);
```

Comment:
Log entry appends are easy to aggregate. If by the time the finisher runs there is >1 entry ready to append, we append them all at once. Since the log entries are always (except when we wrap around the ring) adjacent, we can flush/replicate all of them with a single call and a single RTT for the RDMA. We also update the head pointer and publish all the allocations in a single pmemobj transaction. We have to hold a lock while doing that to synchronize with other appenders, or the flusher.

As implemented we do each pmem buffer flush sequentially in flush_pmem_buffer(), waiting for each RDMA to complete before starting the next. Ideally that interface would take an SGL and do them all in parallel. It's TBD whether an SGL interface is likely to appear. If not, it makes more sense to schedule these individually to a workq that might have >1 thread.